src/nvapi_gpu.cpp

```diff
@@ -260,6 +260,108 @@ extern "C" {
         }
     }
 
+    NvAPI_Status __cdecl NvAPI_GPU_GetCurrentPstate(NvPhysicalGpuHandle hPhysicalGpu, NV_GPU_PERF_PSTATE_ID *pCurrentPstate) {
+        constexpr auto n = "NvAPI_GPU_GetCurrentPstate";
+        static bool alreadyLoggedNoNvml = false;
+        static bool alreadyLoggedHandleInvalidated = false;
+        static bool alreadyLoggedOk = false;
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pCurrentPstate == nullptr)
+            return InvalidArgument(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        if (!adapter->HasNvml())
+            return NoImplementation(n, alreadyLoggedNoNvml);
+
+        if (!adapter->HasNvmlDevice())
+            return HandleInvalidated(str::format(n, ": NVML available but current adapter is not NVML compatible"), alreadyLoggedHandleInvalidated);
+
+        nvmlPstates_t pState;
+        auto result = adapter->GetNvmlPerformanceState(&pState);
+        *pCurrentPstate = static_cast<NV_GPU_PERF_PSTATE_ID>(result);
+
+        return Ok(n, alreadyLoggedOk);
+    }
+
+    NvAPI_Status __cdecl NvAPI_GPU_GetPstates20(NvPhysicalGpuHandle hPhysicalGpu, NV_GPU_PERF_PSTATES20_INFO *pPstatesInfo) {
+        constexpr auto n = "NvAPI_GPU_GetPstates20";
+        static bool alreadyLoggedNoNvml = false;
+        static bool alreadyLoggedHandleInvalidated = false;
+        static bool alreadyLoggedOk = false;
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pPstatesInfo == nullptr)
+            return InvalidArgument(n);
+
+        if (pPstatesInfo->version != NV_GPU_PERF_PSTATES20_INFO_VER && pPstatesInfo->version != NV_GPU_PERF_PSTATES20_INFO_VER2 && pPstatesInfo->version != NV_GPU_PERF_PSTATES20_INFO_VER1)
+            return IncompatibleStructVersion(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        if (!adapter->HasNvml())
+            return NoImplementation(n, alreadyLoggedNoNvml);
+
+        if (!adapter->HasNvmlDevice())
+            return HandleInvalidated(str::format(n, ": NVML available but current adapter is not NVML compatible"), alreadyLoggedHandleInvalidated);
+
+        nvmlPstates_t pState;
+        auto result = adapter->GetNvmlPerformanceState(&pState);
+
+        pPstatesInfo->numPstates = 1;                                       /* TODO: Get number of pstates total */
+        pPstatesInfo->numClocks = 2;                                        /* GPU and Mem clocks */
+        pPstatesInfo->numBaseVoltages = 0;                                  /* nvml does not get voltages */
+        pPstatesInfo->pstates[0].pstateId = static_cast<NV_GPU_PERF_PSTATE_ID>(result);
```

Comment:
This is not result but one of these: https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__gpupstate.html#ga8f33f5167a96a722b885a705520aa25b

Filling the NVAPI `pStatesInfo` structure from the NVML `pState`structure doesn´t work. The former want to have all performance states, the latter only contains the current one.

Reply:
The thing is that as you mention above, there does not seem to be any easy way to build that structure from information from nvml, thus "faking" only having 1 numPstate seemed reasonable - Basically forcing only having Pstate=0 all the time.

Ofc. casting from `result` is fruitless.. will work on that, cos that is obviously the wrong thing to do. 

So, making a "proper" table that are supposed to be in nvapi for this function, i think is not going to work very well. Doing what i did with the table (might just aswell put `pPstatesInfo->pstates[0].pstateId = 0`  (or `NVAPI_GPU_PERF_PSTATE_P0`) is not going to make much difference, cos i do not think it will be "proper" anyway.

Reply:
Using `NVAPI_GPU_PERF_PSTATE_UNDEFINED` instead of `NVAPI_GPU_PERF_PSTATE_P0` might be the most sensible thing here, but it would be useless for applications since they cannot map the current state on it to get the frequencies etc. We could also always return `NVAPI_GPU_PERF_PSTATE_UNDEFINED`as current pstate but then it would be a lottery if you see the correct frequenties or frequencies from other pstates than the current.
I think there is no good way to solve this using nvml.